File: topos/functors/probes/ast/complexity.py
"""
Per-Function Complexity Analysis
--------------------------------

Provides a function-level breakdown of complexity using the AST.
This is separate from the CFG-based module-level cyclomatic complexity used
in the main program evaluation (SIMPLE generator).

Every ``ProgramObject`` produced by a real parse (``ProgramMorphism`` /
``parse_source``) carries a populated, language-neutral ``uast_root``
(see ``topos.graphs.ast.providers.tree_sitter_provider``), so the primary
implementation below walks that UAST directly to find ``FunctionDecl`` /
``MethodDecl`` nodes -- this works uniformly across every language with a
UAST mapper (Python, Rust, Go, JavaScript, TypeScript, C++), unlike the
previous approach of re-querying Python-specific tree-sitter native node
type strings (``function_definition``/``async_function_definition``),
which silently found zero functions for every other language.

A ``uast_root``-less native tree-sitter fallback is retained for callers
that construct a ``ProgramObject`` directly without populating
``uast_root`` (e.g. lightweight test fixtures); it remains Python-only,
same as before this fix.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

from topos.core.object import ProgramObject

if TYPE_CHECKING:
    from topos.graphs.uast.models import UASTNode

_FUNCTION_NODE_TYPES = ("function_definition", "async_function_definition")
_SCOPE_NODE_TYPES = (*_FUNCTION_NODE_TYPES, "class_definition")

DECISION_NODE_TYPES = frozenset(
    {
        "if_statement",
        "elif_clause",
        "for_statement",
        "while_statement",
        "except_clause",
        "with_statement",
        "assert_statement",
        "conditional_expression",  # ternary
        "boolean_operator",  # and/or short-circuit
        "match_statement",  # Python 3.10+
        "case_clause",
        "list_comprehension",
        "dictionary_comprehension",
        "set_comprehension",
        "generator_expression",
    }
)

DECISION_UAST_KINDS = frozenset(
    {
        "IfStmt",
        "ForStmt",
        "WhileStmt",
        "MatchStmt",
        "TryStmt",
    }
)

# UAST kinds that every language mapper agrees denote a callable.
_FUNCTION_UAST_KINDS = frozenset({"FunctionDecl", "MethodDecl"})

# UAST kinds that introduce a new "scope" for qualified-name/kind purposes
# (mirrors ``_SCOPE_NODE_TYPES`` for the native tree-sitter fallback).
_SCOPE_UAST_KINDS = frozenset({*_FUNCTION_UAST_KINDS, "TypeDecl"})

# tree-sitter grammars use different native node kinds for a declaration's
# name token depending on syntactic position (e.g. tree-sitter-javascript
# emits ``property_identifier`` for a class method's name, not
# ``identifier``; tree-sitter-go emits ``field_identifier``/
# ``type_identifier`` elsewhere). ``UASTNode`` has no stored ``name``
# attribute, so recovering it means scanning a declaration's own children
# for one of these native kinds and slicing the corresponding byte span out
# of the source -- the *native* provenance kind is matched here rather than
# the normalized UAST ``kind``, since none of the mappers classify these
# tokens as ``"Identifier"`` today.
_NAME_NATIVE_KINDS = frozenset(
    {
        "identifier",
        "property_identifier",
        "field_identifier",
        "type_identifier",
        "package_identifier",
    }
)

_ASYNC_PREFIX_RE = re.compile(rb"\Aasync\b")


def _calculate_node_complexity(ast: ProgramObject) -> int:
    """Internal helper to calculate complexity of a specific node/sub-tree."""
    if ast.uast_root is not None:
        return _calculate_cyclomatic_complexity_uast(ast.uast_root)

    decision_count = 0

    for node in ast.traverse():
        if node.type in DECISION_NODE_TYPES:
            decision_count += 1
            if node.type == "boolean_operator":
                decision_count += _count_boolean_operators(node) - 1

    return decision_count + 1


def _calculate_cyclomatic_complexity_uast(root: UASTNode) -> int:
    decision_count = 0

    def walk(node) -> None:
        nonlocal decision_count
        kind = getattr(node, "kind", "")
        if kind in DECISION_UAST_KINDS:
            decision_count += 1
        if kind == "BinaryExpr":
            operator = getattr(node, "attributes", {}).get("operator")
            if operator in {"and", "or", "&&", "||"}:
                decision_count += 1
        for child in getattr(node, "children", []):
            walk(child)

    walk(root)
    return decision_count + 1


def _count_boolean_operators(node) -> int:
    """Count chained boolean operators (and/or)."""
    count = 1
    for child in node.children:
        if child.type == "boolean_operator":
            count += _count_boolean_operators(child)
    return count


def _extract_function_name(node, source_bytes: bytes) -> str | None:
    for child in node.children:
        if child.type == "identifier":
            return source_bytes[child.start_byte : child.end_byte].decode("utf-8")
    return None


# ---------------------------------------------------------------------------
# UAST-driven (language-neutral) implementation
# ---------------------------------------------------------------------------


def _extract_uast_name(node: UASTNode, source_bytes: bytes) -> str | None:
    """Best-effort identifier extraction for a UAST declaration node.

    Returns ``None`` for callables with no bound identifier of their own
    (e.g. a JS/TS arrow function assigned via ``const f = (x) => {...}``,
    where the name lives on the enclosing ``variable_declarator``, not the
    ``FunctionDecl`` node itself) -- these are skipped by callers, matching
    the pre-existing behavior of never picking up anonymous callables
    (Python lambdas were never matched by the native node-type query
    either).
    """
    for child in node.children:
        if child.native.node_kind in _NAME_NATIVE_KINDS:
            return source_bytes[child.span.start_byte : child.span.end_byte].decode(
                "utf-8", errors="replace"
            )
    return None


def _is_async_uast(node: UASTNode, source_bytes: bytes) -> bool:
    """Detect a leading ``async`` modifier on a UAST declaration node.

    tree-sitter models ``async`` as an unnamed leading token, so it is
    filtered out of the UAST tree entirely (only *named* children survive
    mapping) -- but it remains part of the declaration node's own byte span,
    so a text sniff at the start of that span recovers it uniformly across
    Python (``async def``), JS/TS (``async function`` / ``async () =>`` /
    ``async foo()``), and Rust (``async fn``). Go has no ``async`` keyword.
    """
    start = node.span.start_byte
    return bool(_ASYNC_PREFIX_RE.match(source_bytes[start : start + 6]))


def _classify_uast_kind(
    own_kind: str, scope_chain: list[tuple[str, str]], is_async: bool
) -> str:
    if own_kind == "MethodDecl":
        return "method"
    if scope_chain:
        enclosing_kind = scope_chain[-1][0]
        if enclosing_kind == "TypeDecl":
            return "method"
        if enclosing_kind in _FUNCTION_UAST_KINDS:
            return "closure"
    return "async_function" if is_async else "function"


@dataclass(frozen=True)
class _UastFunctionEntry:
    node: UASTNode
    name: str
    qualified_name: str
    kind: str


def _iter_uast_function_entries(
    root: UASTNode, source_bytes: bytes
) -> list[_UastFunctionEntry]:
    """Top-down walk of the UAST collecting every named callable, in
    document order, with its dotted qualified name and method/closure/
    function/async_function classification.

    ``UASTNode`` has no parent pointer, so the enclosing-scope chain is
    threaded down explicitly as the recursion descends (mirroring the
    native fallback's upward ``.parent`` walk, just inverted).
    """
    entries: list[_UastFunctionEntry] = []

    def walk(node: UASTNode, scope_chain: list[tuple[str, str]]) -> None:
        kind = getattr(node, "kind", "")
        next_scope_chain = scope_chain

        if kind in _FUNCTION_UAST_KINDS:
            name = _extract_uast_name(node, source_bytes)
            if name is not None:
                qualified = ".".join([n for _, n in scope_chain] + [name])
                is_async = _is_async_uast(node, source_bytes)
                entries.append(
                    _UastFunctionEntry(
                        node=node,
                        name=name,
                        qualified_name=qualified,
                        kind=_classify_uast_kind(kind, scope_chain, is_async),
                    )
                )
            next_scope_chain = [*scope_chain, (kind, name or "<anon>")]
        elif kind in _SCOPE_UAST_KINDS:  # TypeDecl
            name = _extract_uast_name(node, source_bytes)
            next_scope_chain = [*scope_chain, (kind, name or "<anon>")]

        for child in getattr(node, "children", []):
            walk(child, next_scope_chain)

    walk(root, [])
    return entries


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def calculate_function_complexities(ast: ProgramObject) -> dict[str, int]:
    """
    Calculate cyclomatic complexity for each function in the AST.

    Args:
        ast: The ProgramObject to analyze.

    Returns:
        A dictionary mapping function names to their complexity scores.
    """
    if ast.uast_root is not None:
        source_bytes = ast.source.encode("utf-8")
        return {
            entry.name: _calculate_cyclomatic_complexity_uast(entry.node)
            for entry in _iter_uast_function_entries(ast.uast_root, source_bytes)
        }

    # Legacy fallback for ProgramObjects constructed without a uast_root.
    # Python-only, same as before this fix; unreachable for any ProgramObject
    # produced by ProgramMorphism / parse_source, which always populate
    # uast_root for every supported language.
    complexities: dict[str, int] = {}
    source_bytes = ast.source.encode("utf-8")

    for node in ast.nodes_of_type("function_definition", "async_function_definition"):
        func_name = _extract_function_name(node, source_bytes)
        if func_name is None:
            continue

        func_ast = ProgramObject(
            root=node,
            source=ast.source,
            language=ast.language,
        )
        complexities[func_name] = _calculate_node_complexity(func_ast)

    return complexities


def calculate_max_function_complexity(ast: ProgramObject) -> int:
    """Calculate the maximum cyclomatic complexity found in any function."""
    complexities = calculate_function_complexities(ast)
    return max(complexities.values()) if complexities else 0


@dataclass(frozen=True)
class FunctionComplexity:
    """A single function's complexity with its source location and scope kind.

    Drives ``ast.max_function_complexity`` location reporting: ``complexity`` is
    computed with the exact same decision-node logic as
    ``calculate_function_complexities`` (so the max here equals the gate metric),
    and the span/kind let agents map a failing gate back to a concrete edit
    target. ``includes_nested`` is always ``True`` because the AST count walks
    the whole function subtree, including nested callables.
    """

    name: str
    qualified_name: str
    kind: str
    start_line: int
    end_line: int
    complexity: int
    includes_nested: bool = True


def _scope_chain(node, source_bytes: bytes) -> list[tuple[str, str]]:
    """Enclosing ``(type, name)`` scopes, outermost first (excludes ``node``)."""
    chain: list[tuple[str, str]] = []
    parent = node.parent
    while parent is not None:
        if parent.type in _SCOPE_NODE_TYPES:
            chain.append(
                (parent.type, _extract_function_name(parent, source_bytes) or "<anon>")
            )
        parent = parent.parent
    chain.reverse()
    return chain


def _is_async(node) -> bool:
    # tree-sitter-python models ``async def`` as a ``function_definition`` with
    # a leading ``async`` child token (not a distinct node type).
    return node.type == "async_function_definition" or any(
        child.type == "async" for child in node.children
    )


def _classify_kind(node, scope_chain: list[tuple[str, str]]) -> str:
    if scope_chain:
        enclosing_type = scope_chain[-1][0]
        if enclosing_type == "class_definition":
            return "method"
        if enclosing_type in _FUNCTION_NODE_TYPES:
            return "closure"
    return "async_function" if _is_async(node) else "function"


def calculate_function_complexity_entries(
    ast: ProgramObject,
) -> list[FunctionComplexity]:
    """Per-function complexity with locations, parallel to the gate metric.

    Same decision-node counting as ``calculate_function_complexities`` but
    retains every callable (including same-named ones) plus its span, dotted
    qualified name, and scope kind.
    """
    if ast.uast_root is not None:
        source_bytes = ast.source.encode("utf-8")
        return [
            FunctionComplexity(
                name=entry.name,
                qualified_name=entry.qualified_name,
                kind=entry.kind,
                start_line=entry.node.span.start_line,
                end_line=entry.node.span.end_line,
                complexity=_calculate_cyclomatic_complexity_uast(entry.node),
            )
            for entry in _iter_uast_function_entries(ast.uast_root, source_bytes)
        ]

    # Legacy fallback for ProgramObjects constructed without a uast_root.
    # Python-only, same as before this fix; see calculate_function_complexities.
    source_bytes = ast.source.encode("utf-8")
    entries: list[FunctionComplexity] = []

    for node in ast.nodes_of_type(*_FUNCTION_NODE_TYPES):
        name = _extract_function_name(node, source_bytes)
        if name is None:
            continue
        scope_chain = _scope_chain(node, source_bytes)
        qualified = ".".join([n for _, n in scope_chain] + [name])
        func_ast = ProgramObject(root=node, source=ast.source, language=ast.language)
        entries.append(
            FunctionComplexity(
                name=name,
                qualified_name=qualified,
                kind=_classify_kind(node, scope_chain),
                start_line=node.start_point[0] + 1,
                end_line=node.end_point[0] + 1,
                complexity=_calculate_node_complexity(func_ast),
            )
        )

    return entries

File: crates/topos-core/src/functors/probes/ast/complexity.rs
//! Per-function complexity analysis over the UAST.
//!
//! # Deviation from the Python original
//!
//! Python's `calculate_function_complexities` builds a per-function
//! sub-`ProgramObject` without wiring `uast_root`, so its intended
//! language-neutral path (walking `DECISION_UAST_KINDS`) never actually
//! runs — it silently falls back to Python-specific native tree-sitter
//! node-type strings (`"function_definition"`, `"elif_clause"`, ...),
//! so every other language gets `max_function_complexity = 0` always (a
//! vacuous pass of the `<= 10.0` gate). Filed as issue #153.
//!
//! This is a from-scratch implementation rather than a faithful port of
//! that bug: it walks the already-built UAST directly for
//! `FunctionDecl`/`MethodDecl` nodes, genuinely multi-language, and
//! simpler than Python's per-function AST reconstruction.

use std::collections::HashMap;

use crate::graphs::uast::models::{AttributeValue, UASTNode};

const DECISION_UAST_KINDS: &[&str] = &["IfStmt", "ForStmt", "WhileStmt", "TryStmt"];

/// Cyclomatic complexity of one callable's subtree: each decision node
/// (`IfStmt`/`ForStmt`/`WhileStmt`/`TryStmt`) adds one, a `MatchStmt` adds
/// one per case arm beyond the first (a k-way switch is k-1 decisions), so
/// this agrees with `cfg.cyclomatic`; plus a short-circuit `BinaryExpr`
/// (`and`/`or`/`&&`/`||`) adds one.
///
/// Note: counting per-arm here intentionally diverges from the last Python
/// release (`topos-mcp==0.3.11`), which counted a whole match/switch as a
/// single decision. The divergence is documented in the `[0.4.0]` CHANGELOG
/// entry (the parity/benchmark harness that originally allowlisted it was a
/// migration-verification artifact and has since been removed).
///
/// The boolean-operator check is currently dormant — no UAST mapper
/// (issue #142) populates a `BinaryExpr`'s `"operator"` attribute with
/// token text yet, the same "mappers don't carry token text" limitation
/// noted in `graphs::pdg::object::identifier_name`. Kept because it's
/// what the Python original's own (also dormant) UAST-kind path
/// intended, and it costs nothing to leave wired for when a mapper
/// starts recording operator text.
fn node_complexity(node: &UASTNode) -> usize {
    fn walk(node: &UASTNode, count: &mut usize) {
        match node.kind.as_str() {
            // A k-way switch/match contributes k branches (k - 1 decisions),
            // counted from its arms so this agrees with the CFG builder.
            "MatchStmt" => {
                *count += crate::graphs::cfg::builder::match_arm_count(node).saturating_sub(1);
            }
            k if DECISION_UAST_KINDS.contains(&k) => *count += 1,
            "BinaryExpr" => {
                if let Some(AttributeValue::Str(op)) = node.attributes.get("operator") {
                    if matches!(op.as_str(), "and" | "or" | "&&" | "||") {
                        *count += 1;
                    }
                }
            }
            _ => {}
        }
        for child in &node.children {
            walk(child, count);
        }
    }
    let mut count = 0;
    walk(node, &mut count);
    count + 1
}

/// Cyclomatic complexity for each function/method in a UAST, keyed by
/// UAST node id.
///
/// Python keys by extracted function *name*; this crate's mappers don't
/// carry token text yet (same limitation as `pdg::object`), so node id
/// is the only stable key available — and it has the incidental benefit
/// of not colliding on same-named nested/overloaded functions the way a
/// name-keyed map would.
pub fn calculate_function_complexities(uast_root: &UASTNode) -> HashMap<String, usize> {
    let mut complexities = HashMap::new();
    let mut stack: Vec<&UASTNode> = vec![uast_root];
    while let Some(node) = stack.pop() {
        if matches!(node.kind.as_str(), "FunctionDecl" | "MethodDecl") {
            complexities.insert(node.id.clone(), node_complexity(node));
        }
        stack.extend(node.children.iter());
    }
    complexities
}

/// Maximum cyclomatic complexity found in any function/method; `0` if
/// there are none.
pub fn calculate_max_function_complexity(uast_root: &UASTNode) -> usize {
    calculate_function_complexities(uast_root)
        .values()
        .copied()
        .max()
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graphs::ast::dispatch::parse_source;

    #[test]
    fn flat_function_has_complexity_one() {
        let result = parse_source("def f(x):\n    return x\n", "python", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 1);
    }

    #[test]
    fn nested_if_and_for_increase_complexity() {
        let source = "def f(items):\n    for item in items:\n        if item:\n            return item\n    return None\n";
        let result = parse_source(source, "python", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 3);
    }

    #[test]
    fn works_across_languages_unlike_the_python_original() {
        // The point of issue #153: Rust functions must be counted too.
        let source = "fn f(x: i32) -> i32 {\n    if x > 0 {\n        return x;\n    }\n    0\n}\n";
        let result = parse_source(source, "rust", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 2);
    }

    #[test]
    fn no_functions_is_zero() {
        let result = parse_source("x = 1\n", "python", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 0);
    }

    #[test]
    fn python_match_arms_count_toward_complexity() {
        // Each case arm is a decision: a 3-arm match => 2 decisions + base 1
        // = complexity 3, consistent with cfg.cyclomatic (#153 follow-up).
        // Intentionally diverges from 0.3.11 (allowlisted in parity_check.py).
        let source = "def f(x):\n    match x:\n        case 1:\n            y = 1\n        case 2:\n            y = 2\n        case _:\n            y = 3\n    return y\n";
        let result = parse_source(source, "python", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 3);
    }

    #[test]
    fn go_switch_arms_count_toward_complexity() {
        let source = "package p\nfunc f(x int) int {\n\tvar y int\n\tswitch {\n\tcase x > 2:\n\t\ty = 1\n\tcase x > 1:\n\t\ty = 2\n\tdefault:\n\t\ty = 3\n\t}\n\treturn y\n}\n";
        let result = parse_source(source, "go", None).unwrap();
        assert_eq!(calculate_max_function_complexity(&result.uast_root), 3);
    }
}

// ---------------------------------------------------------------------------
// Per-function complexity *entries* — name/span-aware, still UAST-only
// ---------------------------------------------------------------------------
//
// `calculate_function_complexities` above answers "what's the worst
// complexity" (keyed by opaque node id, fine for a gate check).
// `calculate_function_complexity_entries` answers "which function, at
// which lines" for agent-facing reporting — it needs a real name and a
// span. Both requirements are satisfiable straight from the UAST: UAST
// spans already carry real line numbers, and a `FunctionDecl`'s name is
// its first `Identifier`-kind child (the mappers preserve that child;
// they just don't duplicate its text into an attribute) — so this reuses
// `node_complexity` above rather than re-deriving complexity via a
// second, tree-sitter-native pass. Genuinely multi-language, same as
// `calculate_function_complexities`.

const SCOPE_UAST_KINDS: &[&str] = &["FunctionDecl", "MethodDecl", "TypeDecl"];

/// One function/method/closure's complexity, name, span, and scope kind.
pub struct FunctionComplexityEntry {
    pub name: String,
    pub qualified_name: String,
    pub kind: &'static str,
    pub start_line: usize,
    pub end_line: usize,
    pub complexity: usize,
}

/// A UAST node's own name, if its first child is an `Identifier` (the
/// mappers place the declared name there for `FunctionDecl` /
/// `MethodDecl` / `TypeDecl`). Sliced from `source` by the child's span,
/// since UAST nodes don't carry token text themselves.
fn uast_node_name(node: &UASTNode, source: &str) -> Option<String> {
    let ident = node.children.first()?;
    if ident.kind != "Identifier" {
        return None;
    }
    source
        .get(ident.span.start_byte..ident.span.end_byte)
        .map(|s| s.to_string())
}

fn classify_kind(node: &UASTNode, source: &str, chain: &[(String, String)]) -> &'static str {
    if let Some((enclosing_kind, _)) = chain.last() {
        if enclosing_kind == "TypeDecl" {
            return "method";
        }
        if enclosing_kind == "FunctionDecl" || enclosing_kind == "MethodDecl" {
            return "closure";
        }
    }
    if node.kind == "MethodDecl" {
        return "method";
    }
    if is_async(node, source) {
        "async_function"
    } else {
        "function"
    }
}

/// Best-effort `async` detection: the mappers only keep *named* tree-sitter
/// children (see `mapper_common::filtered_named_children`), and `async` is
/// an anonymous keyword token in tree-sitter-python's grammar — the node
/// kind stays `function_definition` either way, but its *span* still
/// starts at `async` (tree-sitter includes leading anonymous tokens in the
/// parent's span), so it never survives as a UAST child. Checking whether
/// the node's own span starts with the `async` keyword recovers it without
/// touching the shared mapper.
fn is_async(node: &UASTNode, source: &str) -> bool {
    source.get(node.span.start_byte..).is_some_and(|text| {
        text.strip_prefix("async")
            .is_some_and(|rest| rest.starts_with(char::is_whitespace))
    })
}

fn collect_entries(
    node: &UASTNode,
    source: &str,
    chain: &mut Vec<(String, String)>,
    entries: &mut Vec<FunctionComplexityEntry>,
) {
    let is_function = matches!(node.kind.as_str(), "FunctionDecl" | "MethodDecl");
    let is_scope = SCOPE_UAST_KINDS.contains(&node.kind.as_str());

    if is_function {
        if let Some(name) = uast_node_name(node, source) {
            let mut qualified_parts: Vec<&str> = chain.iter().map(|(_, n)| n.as_str()).collect();
            qualified_parts.push(&name);
            entries.push(FunctionComplexityEntry {
                name: name.clone(),
                qualified_name: qualified_parts.join("."),
                kind: classify_kind(node, source, chain),
                start_line: node.span.start_line,
                end_line: node.span.end_line,
                complexity: node_complexity(node),
            });
        }
    }

    let pushed = if is_scope {
        uast_node_name(node, source).map(|name| {
            chain.push((node.kind.clone(), name));
        })
    } else {
        None
    };

    for child in &node.children {
        collect_entries(child, source, chain, entries);
    }

    if pushed.is_some() {
        chain.pop();
    }
}

/// Per-function complexity with locations, parallel to the gate metric.
///
/// Same decision-node counting as [`calculate_function_complexities`],
/// but keyed by real (dotted, qualified) names with spans. `source` is
/// needed to slice out identifier text (see [`uast_node_name`]).
pub fn calculate_function_complexity_entries(
    uast_root: &UASTNode,
    source: &str,
) -> Vec<FunctionComplexityEntry> {
    let mut entries = Vec::new();
    let mut chain = Vec::new();
    collect_entries(uast_root, source, &mut chain, &mut entries);
    entries
}

#[cfg(test)]
mod entries_tests {
    use super::*;
    use crate::graphs::ast::dispatch::parse_source;

    fn entries(source: &str, language: &str) -> Vec<FunctionComplexityEntry> {
        let result = parse_source(source, language, None).expect("parse should not fail");
        calculate_function_complexity_entries(&result.uast_root, source)
    }

    #[test]
    fn top_level_function_kind_and_span() {
        let source = "def foo(x):\n    if x:\n        return 1\n    return 0\n";
        let es = entries(source, "python");
        assert_eq!(es.len(), 1);
        let foo = &es[0];
        assert_eq!(foo.name, "foo");
        assert_eq!(foo.qualified_name, "foo");
        assert_eq!(foo.kind, "function");
        assert_eq!(foo.start_line, 1);
        assert_eq!(foo.complexity, 2);
    }

    #[test]
    fn method_inside_class_is_qualified_and_kind_method() {
        let source = "class C:\n    def m(self):\n        return 1\n";
        let es = entries(source, "python");
        let m = es.iter().find(|e| e.name == "m").unwrap();
        assert_eq!(m.qualified_name, "C.m");
        assert_eq!(m.kind, "method");
    }

    #[test]
    fn nested_closure_is_dotted_and_outer_includes_nested_count() {
        let source = "def outer():\n    def inner():\n        if True:\n            return 1\n    return inner\n";
        let es = entries(source, "python");
        let inner = es.iter().find(|e| e.name == "inner").unwrap();
        let outer = es.iter().find(|e| e.name == "outer").unwrap();
        assert_eq!(inner.qualified_name, "outer.inner");
        assert_eq!(inner.kind, "closure");
        assert!(outer.complexity >= inner.complexity);
    }

    #[test]
    fn module_level_only_has_no_entries() {
        assert!(entries("x = 1\n", "python").is_empty());
    }

    #[test]
    fn async_function_kind_is_detected() {
        let es = entries("async def bar():\n    return 1\n", "python");
        assert_eq!(es[0].kind, "async_function");
    }

    #[test]
    fn works_across_languages_unlike_the_python_original() {
        let source = "fn f(x: i32) -> i32 {\n    if x > 0 {\n        return x;\n    }\n    0\n}\n";
        let es = entries(source, "rust");
        assert_eq!(es.len(), 1);
        assert_eq!(es[0].name, "f");
        assert_eq!(es[0].complexity, 2);
    }
}
